### PR TITLE
Make all properties in ApiResource nullable with null as default value

### PR DESCRIPTION
**Pull request by Zowac**
| Q             | A
| ------------- | ---
| Branch?       | main
| Tickets       | #4411 
| License       | MIT

Make sure that the ApiResource properties are nullable and have null by default

### CODE REVIEW

File: src/Metadata/ApiResource.php [hunk 1]
```diff
@@ -215,7 +215,7 @@ public function __construct(
         array $exceptionToStatus = [],
         ?bool $queryParameterValidationEnabled = null,
         ?array $graphQlOperations = null,
-        array $extraProperties = []
+        ?array $extraProperties = []
```
Comment (soyuka) [inline src/Metadata/ApiResource.php:218]:
```suggestion
        array $extraProperties = []
```

File: src/Metadata/ApiResource.php [hunk 2]
```diff
@@ -330,8 +330,12 @@ public function withDescription(?string $description = null): self
         return $self;
     }
 
-    public function getTypes(): array
+    public function getTypes(): ?array
     {
+        if (!$this->types) {
+            $this->types = [$this->shortName];
+        }
+
```
Comment (soyuka) [inline src/Metadata/ApiResource.php:338]: revert this there should not be any logic here

Comment (Zowac) [reply src/Metadata/ApiResource.php:338]: Oups sorry it wasn't supposed to be pushed ^^'

**Comment by dunglas** [comment]
Is this included in https://github.com/api-platform/core/pull/4422?

File: src/Core/Serializer/SerializerContextBuilder.php [hunk 3]
```diff
@@ -66,7 +66,7 @@ public function createFromRequest(Request $request, bool $normalization, array $
             $context['iri_only'] = $context['iri_only'] ?? false;
             $context['request_uri'] = $request->getRequestUri();
             $context['uri'] = $request->getUri();
-            $context['input'] = $operation->getInput();
+            $context['input'] = $operation->getInput() ?? null;
```
Comment (soyuka) [inline src/Core/Serializer/SerializerContextBuilder.php:69]: revert

File: src/GraphQl/Type/TypeBuilder.php [hunk 4]
```diff
@@ -107,7 +107,7 @@ public function getResourceObjectType(?string $resourceClass, ResourceMetadataCo
             return $resourceObjectType;
         }
 
-        $ioMetadata = $input ? $operation->getInput() : $operation->getOutput();
+        $ioMetadata = $input ? $operation->getInput() ?? null : $operation->getOutput() ?? null;
```
Comment (soyuka) [inline src/GraphQl/Type/TypeBuilder.php:110]: revert

Comment (Zowac) [reply src/GraphQl/Type/TypeBuilder.php]: done

File: src/Core/Metadata/Resource/ApiResourceToLegacyResourceMetadataTrait.php [hunk 5]
```diff
@@ -57,7 +57,11 @@ private function transformResourceToResourceMetadata(ApiResource $resource): Res
             $graphqlOperations[$operationName] = $this->toArray($operation);
         }
 
-        return new ResourceMetadata($resource->getShortName(), $resource->getDescription(), $resource->getTypes()[0] ?? null, $itemOperations, $collectionOperations, $attributes, null, $graphqlOperations);
+        if (!$type = $resource->getTypes()) {
+            $type = [$resource->getShortName()];
+        }
+
+        return new ResourceMetadata($resource->getShortName(), $resource->getDescription(), $type[0] ?? null, $itemOperations, $collectionOperations, $attributes, null, $graphqlOperations);
```
Comment (soyuka) [inline src/Core/Metadata/Resource/ApiResourceToLegacyResourceMetadataTrait.php:64]:
```suggestion
        return new ResourceMetadata($resource->getShortName(), $resource->getDescription(), $resource->getTypes()[0] ?? null, $itemOperations, $collectionOperations, $attributes, null, $graphqlOperations);
```

**Comment by soyuka** [comment]
superseeded by https://github.com/api-platform/core/pull/4449 thanks @Zowac !

**Pull request closed without merging**